Framework/Geometry/src/Instrument/ParameterMap.cpp

```diff
@@ -1161,24 +1164,54 @@ bool ParameterMap::hasDetectorInfo(const Instrument *instrument) const {
     return false;
   return static_cast<bool>(m_detectorInfo);
 }
+
+/** Only for use by ExperimentInfo. Returns returns true if this instrument
+ contains a ComponentInfo.
+
+ The `instrument` argument is needed for the special case of having a neutronic
+ *and* a physical instrument. `Instrument` uses the same parameter map for both,
+ but the ComponentInfo is only for the neutronic instrument. */
+bool ParameterMap::hasComponentInfo(const Instrument *instrument) const {
+  if (instrument != m_instrument)
+    return false;
+  return static_cast<bool>(m_componentInfo);
+}
+
 /// Only for use by ExperimentInfo. Returns a reference to the DetectorInfo.
 const Beamline::DetectorInfo &ParameterMap::detectorInfo() const {
   if (!hasDetectorInfo(m_instrument))
     throw std::runtime_error("Cannot return reference to NULL DetectorInfo");
   return *m_detectorInfo;
 }
 
+/// Only for use by ExperimentInfo. Returns a reference to the ComponentInfo.
+const Beamline::ComponentInfo &ParameterMap::componentInfo() const {
+  if (!hasComponentInfo(m_instrument)) {
+    throw std::runtime_error("Cannot return reference to NULL ComponentInfo");
+  }
+  return *m_componentInfo;
+}
+
 /// Only for use by Detector. Returns a detector index for a detector ID.
 size_t ParameterMap::detectorIndex(const detid_t detID) const {
   return m_instrument->detectorIndex(detID);
 }
 
+size_t ParameterMap::componentIndex(const ComponentID componentId) const {
+  return m_instrument->componentIndex(componentId);
```

Comment:
@OwenArnold This is what I meant on slack: Replace this line by
```cpp
return m_componentInfo->indexOf(componentId);
```
With that change a fair number of unit tests seems to pass (still many of them are segfaulting though).

All you need to do is move `API::ComponentInfo` into `Geometry`, I think (I just fudged it for my test by making `API::ComponentInfo.h` header only so I could include it in `Geometry` files).